ISPN-12810 marshalling and encoding title (#9106)
infinispan/infinispan: documentation/src/main/asciidoc/topics/proc_registering_sci_remote_caches.adoc

[id='registering-sci-remote-caches_{context}']
= Registering serialization context initializers with {brandname} Server

> You must register your Protobuf schema with {brandname} Server before you can marshall user types to and from remote caches.

Reviewer: That's only true if you wish todo anything on the server side that requires knowledge of the user types protobuf format. If no transformations, i.e. transcoding or marshalling, occur then registration can be skipped as the values are just handled as  a simple `byte[]`.

Reviewer 2: As this is about SerializationContextinitializer it is protobuf so the schema is needed

.Prerequisites

* Generate a `SerializationContextInitializer` implementation with the ProtoStream processor.
+
After the build completes, you can find generated Protobuf schema in the `target/<schemaFilePath>/` directory.

.Procedure

Do one of the following to register your generated Protobuf schema with {brandname} Server:

* Add the generated Protobuf schema to the `\___protobuf_metadata` cache with the {brandname} command line interface (CLI) or REST interface.
+
This method ensures that {brandname} automatically distributes your schema across the cluster.

* Use the generated `SerializationContextInitializer` implementation with a Hot Rod client to register the Protobuf schema, as in the following example:
+
[source,java,options="nowrap",subs=attributes+]
----
include::code_examples/LibraryInitializerClientSchema.java[]
----

> As an alternative, you can add an archive with the `SerializationContextInitializer` implementation and custom classes to the `{server_home}/server/lib` directory.

Reviewer: This is an alternative to remote proto schema registration if only the schemas are required, i.e. for protostream -> json transcoding, however this is an essential step if protostream -> x-java-object transformations are required at any time on the server as the ProtoStream marshallers are required on the server for that type. Similarly the POJO jar(s) must be present.

[id='registering-sci-remote-caches_{context}']

> = Registering serialization context initializers with {brandname} Server

Reviewer 3: ```suggestion
= Registering Protobuf schemas with {brandname} Server
```

> You must register Protobuf schema with {brandname} Server if you want other clients to operate on your custom types, perform Ickle queries, or convert custom types between media types.

Reviewer 3: ```suggestion
You must register Protobuf schema with {brandname} Server if you want to perform Ickle queries, convert from `application/x-protostream` to other media types like `application/json`.
```
I don't think you need the schema on the server for other clients to operate on your custom types, the other clients can get the schema elsewhere (and having just the schema doesn't mean they can unmarshall your POJOs, if they don't have the classes and the generated marshallers).

Author: I've changed this

+
[source,java,options="nowrap",subs=attributes+]
----
include::code_examples/LibraryInitializerClientSchema.java[]
----

* Add a JAR file with the `SerializationContextInitializer` implementation and custom classes to the `{server_home}/server/lib` directory.
+
When you do this, {brandname} Server registers your Protobuf schema at startup. However, you must add the archive to each server installation because the schema are not saved in the `___protobuf_metadata` cache or automatically distributed across the cluster.
+
[NOTE]
====
You must do this if you require {brandname} Server to perform any `application/x-protostream` to `application/x-java-object` conversions, in which case you must also add any JAR files for your POJOs.
====

.Next steps

Register the `SerializationContextInitializer` with your Hot Rod clients, as in the following example:

[source,java,options="nowrap",subs=attributes+]
----
include::code_examples/LibraryInitializerClient.java[]
----

[role="_additional-resources"]
.Additional resources
* link:{cli_docs}#registering_protobuf_schemas[CLI: Registering Protobuf schemas]
* link:{rest_docs}#rest_v2_create_protobuf[REST API: Creating Protobuf schemas]

> * link:{protostreamdocroot}/org/infinispan/protostream/annotations/ProtoAdapter.html[org.infinispan.protostream.annotations.ProtoAdapter]

Author: removed
